OpenRA.Game/Map/MapDirectoryTracker.cs

```diff
@@ -0,0 +1,121 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using OpenRA.FileSystem;
+
+namespace OpenRA
+{
+	public class MapDirectoryTracker
+	{
+		readonly FileSystemWatcher watcher;
+		readonly MapGrid mapGrid;
+		readonly IReadOnlyPackage package;
+		readonly MapClassification classification;
+
+		enum MapAction { Add, Delete, Update }
+		readonly Dictionary<string, MapAction> mapActionQueue = new Dictionary<string, MapAction>();
+
+		bool dirty = false;
+
+		public MapDirectoryTracker(MapGrid mapGrid, IReadOnlyPackage package, MapClassification classification)
+		{
+			this.mapGrid = mapGrid;
+			this.package = package;
+			this.classification = classification;
+
+			watcher = new FileSystemWatcher(package.Name);
+			watcher.Changed += (object sender, FileSystemEventArgs e) => AddMapAction(MapAction.Update, e.FullPath);
+			watcher.Created += (object sender, FileSystemEventArgs e) => AddMapAction(MapAction.Add, e.FullPath);
+			watcher.Deleted += (object sender, FileSystemEventArgs e) => AddMapAction(MapAction.Delete, e.FullPath);
+			watcher.Renamed += (object sender, RenamedEventArgs e) => AddMapAction(MapAction.Add, e.FullPath, e.OldFullPath);
+
+			watcher.IncludeSubdirectories = true;
+			watcher.EnableRaisingEvents = true;
+		}
+
+		void AddMapAction(MapAction mapAction, string fullpath, string oldFullPath = null)
+		{
+			dirty = true;
```

Comment:
The `FileSystemWatcher` is going to raise these callbacks on a different thread, we'll want to lock to avoid threading issues.

Around the entire body of `AddMapAction` and `UpdateMaps` I would add `lock (mapActionQueue) { ... }`.

Reply:
I'm not entirely sure how `lock` works. When I try applying the lock statement throws this error `The name 'mapActionQueue' does not exist in the current context.`. I also appears that I need to include` RemoveSubDirs` in the lock statement

Reply:
e.g.

```c#
void AddMapAction(MapAction mapAction, string fullpath, string oldFullPath = null)
{
    lock (mapActionQueue)
    {
        // body of the AddMapAction method
    }
}
```

Reply:
done